drivers/net/nsos_adapt.c

```diff
@@ -0,0 +1,673 @@
+int nsos_adapt_socket(int family_mid, int type_mid, int proto_mid)
+{
+	int family;
+	int type;
+	int proto;
+	int ret;
+
+	ret = socket_family_from_nsos_mid(family_mid, &family);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = socket_type_from_nsos_mid(type_mid, &type);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = socket_proto_from_nsos_mid(proto_mid, &proto);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = socket(family, type, proto);
+	if (ret < 0) {
+		return -errno_to_nsos_mid(errno);
+	}
+
+	return ret;
+}
+
+static int sockaddr_from_nsos_mid(struct sockaddr **addr, socklen_t *addrlen,
+				  const struct nsos_mid_sockaddr *addr_mid, size_t addrlen_mid)
+{
+	if (!addr_mid) {
+		*addr = NULL;
+		*addrlen = 0;
+
+		return 0;
+	}
+
+	switch (addr_mid->sa_family) {
+	case NSOS_MID_AF_INET: {
+		const struct nsos_mid_sockaddr_in *addr_in_mid =
+			(const struct nsos_mid_sockaddr_in *)addr_mid;
+		struct sockaddr_in *addr_in = (struct sockaddr_in *)*addr;
+
+		addr_in->sin_family = AF_INET;
+		addr_in->sin_port = addr_in_mid->sin_port;
+		addr_in->sin_addr.s_addr = addr_in_mid->sin_addr;
+
+		*addrlen = sizeof(*addr_in);
+
+		return 0;
+	}
+	case NSOS_MID_AF_INET6: {
+		const struct nsos_mid_sockaddr_in6 *addr_in_mid =
+			(const struct nsos_mid_sockaddr_in6 *)addr_mid;
+		struct sockaddr_in6 *addr_in = (struct sockaddr_in6 *)*addr;
+
+		addr_in->sin6_family = AF_INET6;
+		addr_in->sin6_port = addr_in_mid->sin6_port;
+		addr_in->sin6_flowinfo = 0;
+		memcpy(addr_in->sin6_addr.s6_addr, addr_in_mid->sin6_addr,
+		       sizeof(addr_in->sin6_addr.s6_addr));
+		addr_in->sin6_scope_id = addr_in_mid->sin6_scope_id;
+
+		*addrlen = sizeof(*addr_in);
+
+		return 0;
+	}
+	}
+
+	return -NSOS_MID_EINVAL;
+}
+
+static int sockaddr_to_nsos_mid(const struct sockaddr *addr, socklen_t addrlen,
+				struct nsos_mid_sockaddr *addr_mid, size_t *addrlen_mid)
+{
+	if (!addr) {
+		*addrlen_mid = 0;
+
+		return 0;
+	}
+
+	switch (addr->sa_family) {
+	case AF_INET: {
+		struct nsos_mid_sockaddr_in *addr_in_mid =
+			(struct nsos_mid_sockaddr_in *)addr_mid;
+		const struct sockaddr_in *addr_in = (const struct sockaddr_in *)addr;
+
+		if (addr_in_mid) {
+			addr_in_mid->sin_family = NSOS_MID_AF_INET;
+			addr_in_mid->sin_port = addr_in->sin_port;
+			addr_in_mid->sin_addr = addr_in->sin_addr.s_addr;
+		}
+
+		if (addrlen_mid) {
+			*addrlen_mid = sizeof(*addr_in);
+		}
+
+		return 0;
+	}
+	case AF_INET6: {
+		struct nsos_mid_sockaddr_in6 *addr_in_mid =
+			(struct nsos_mid_sockaddr_in6 *)addr_mid;
+		const struct sockaddr_in6 *addr_in = (const struct sockaddr_in6 *)addr;
+
+		if (addr_in_mid) {
+			addr_in_mid->sin6_family = NSOS_MID_AF_INET6;
+			addr_in_mid->sin6_port = addr_in->sin6_port;
+			memcpy(addr_in_mid->sin6_addr, addr_in->sin6_addr.s6_addr,
+			       sizeof(addr_in_mid->sin6_addr));
+			addr_in_mid->sin6_scope_id = addr_in->sin6_scope_id;
+		}
+
+		if (addrlen_mid) {
+			*addrlen_mid = sizeof(*addr_in);
+		}
+
+		return 0;
+	}
+	}
+
+	return -NSOS_MID_EINVAL;
```

Comment:
seems ignored in all calls(?)

Reply:
Updated all callers to take this return value into account.